_source/_posts/2019-11-19-secure-kafka-java.md

```diff
@@ -0,0 +1,458 @@
+
+@Configuration
+public class ConsumerConfiguration {
+
+    private static final String KAFKA_BROKER = "localhost:9092";
+    private static final String GROUP_ID = "kafka-sandbox";
+
+    @Bean
+    public ConsumerFactory<String, String> consumerFactory() {
+        return new DefaultKafkaConsumerFactory<>(consumerConfigurations());
+    }
+
+    @Bean
+    public Map<String, Object> consumerConfigurations() {
+        Map<String, Object> configurations = new HashMap<>();
+
+        configurations.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, KAFKA_BROKER);
+        configurations.put(ConsumerConfig.GROUP_ID_CONFIG, GROUP_ID);
+        configurations.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        configurations.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+
+        return configurations;
+    }
+
+    @Bean
+    ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory() {
+        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+        factory.setConsumerFactory(consumerFactory());
+        return factory;
+    }
+
+}
+```
+
+The code above creates a factory that knows how to connect to your local broker. It also configures your consumer to deserialize a `String` for both the key and the value, matching the producer configuration.
+
+The Group ID is mandatory and used by Kafka to allow parallel data consumption. The `ConcurrentKafkaListenerContainerFactory` bean allows your app to consume messages in more than one thread.
+
+Now that your Java app is configured to find consumers inside your Kafka broker, let's start listening to the messages sent to the topic.
+
+Inside the `src/main/java/com/okta/javakafka/consumer` directory, create the following class:
+
+```java
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+public class MyTopicConsumer {
+
+    private final List<String> messages = new ArrayList<>();
+
+    @KafkaListener(topics = "myTopic", groupId = "kafka-sandbox")
+    public void listen(String message) {
+        synchronized (messages) {
+            messages.add(message);
+        }
+    }
+
+    public List<String> getMessages() {
+        return messages;
+    }
+
+}
+```
+
+This class is responsible for listening to changes inside the `myTopic` topic. It does so by using the `KafkaListener` annotation. Every time a new message is sent from a producer to the topic, your app receives a message inside this class. It adds a message to the list of messages received, making it available to other classes through the `getMessages()` method.
+
+Next, let's create an endpoint that displays a list of consumed messages. Go back to the `KafkaController` to add `MyTopicConsumer` as a dependency and a `getMessages()` endpoint. 
```

Comment:
...and a `getMessages()` method. (currently says: endpoint)

Reply:
Good catch, will fix.